test/integration/dra/binding_conditions_test.go

```diff
@@ -606,3 +557,52 @@ profiles:
 
 	waitForPodScheduled(tCtx, namespace, pod.Name)
 }
+var claimDeviceStatusRetry = wait.Backoff{
+	Steps:    10,
+	Duration: 50 * time.Millisecond,
+	Factor:   1.5,
+	Jitter:   0.1,
+}
+
+// addAllocatedDeviceCondition updates status.devices using the currently allocated device.
+// It retries longer than the default backoff because the scheduler may be updating the same
+// ResourceClaim concurrently when these tests toggle binding conditions.
+func addAllocatedDeviceCondition(tCtx ktesting.TContext, namespace, claimName string, condition metav1.Condition) {
+	tCtx.Helper()
+
+	err := retry.RetryOnConflict(claimDeviceStatusRetry, func() error {
+		latest, err := tCtx.Client().ResourceV1().ResourceClaims(namespace).Get(tCtx, claimName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if latest.Status.Allocation == nil || len(latest.Status.Allocation.Devices.Results) == 0 {
+			return fmt.Errorf("claim %s has no allocated devices to update status for", claimName)
+		}
+
+		statuses := make([]resourceapi.AllocatedDeviceStatus, len(latest.Status.Allocation.Devices.Results))
+		for i, result := range latest.Status.Allocation.Devices.Results {
+			cond := condition
+			cond.ObservedGeneration = latest.Generation
+			if cond.LastTransitionTime.IsZero() {
+				cond.LastTransitionTime = metav1.Now()
+			}
+			var shareID *string
+			if result.ShareID != nil {
+				s := string(*result.ShareID)
+				shareID = &s
+			}
+			statuses[i] = resourceapi.AllocatedDeviceStatus{
+				Driver:     result.Driver,
+				Pool:       result.Pool,
+				Device:     result.Device,
+				ShareID:    shareID,
+				Conditions: []metav1.Condition{cond},
+			}
+		}
```

Comment:
> Copies scheduler-written fields instead of overwriting them

The scheduler doesn't write these fields, does it?

It's unclear whether `addAllocatedDeviceCondition` should add the condition to all allocated devices. IMHO the individual tests were easier to read without this new helper.